Heap Buffer Overflow in MJPEG Encoder mjpeg_buffer::put_bits via Small Frame (1×1)

### System Information


Field | Value
-- | --
OS | Ubuntu 24.04.4 LTS (Noble Numbat)
Kernel | 7.0.0-15-generic
Architecture | x86_64
CPU | Intel(R) Xeon(R) CPU Max 9468 (2×48c/96t, 192 logical)
RAM | 2.1 TiB
Compiler | Ubuntu clang 21.1.8
CMake | 3.28.3
Sanitizers | address, undefined
Build type | Release + -g -O1 -fsanitize=address,undefined

#29113 tries to fix this.

### Detailed description

## Summary
A heap buffer overflow (WRITE of size 4) occurs in the built-in MJPEG encoder's `mjpeg_buffer::put_bits` function when writing a frame whose dimensions cause the initial per-stripe buffer to be allocated with exactly 2 slots. When a `put_bits` call consumes exactly all remaining bits in the last slot (`len == bits_free`), the resize guard is not triggered, but the function unconditionally writes into the next (out-of-bounds) slot.

## Version
```
$ git describe
4.13.0-150-gc7732e1043
```

## Description

### Buffer Allocation

When `MjpegEncoder` is constructed, each stripe's output buffer is sized as:

```cpp
// cap_mjpeg_encoder.cpp:1200
m_buffer_list.allocate_buffers(stripes_count, (height * width * 2) / stripes_count);
```

For a **1×1 grayscale frame** with `stripes_count = 1`, this yields `(1 * 1 * 2) / 1 = 2` unsigned-int slots (8 bytes total).

### Off-by-One in Resize Guard

The vulnerable function is `put_bits` (line 163–178):

```cpp
// cap_mjpeg_encoder.cpp:163
inline void put_bits(unsigned bits, int len)
{
    CV_Assert(len >= 0 && len < 32);
    if ((m_pos == (data.size() - 1) && len > bits_free) || m_pos == data.size())
    {                                      // ^^^ BUG: should be >=
        resize(int(2 * data.size()));
    }

    bits_free -= (len);
    unsigned int tempval = (bits) & bit_mask[(len)];

    if (bits_free <= 0)
    {
        data[m_pos] |= ((unsigned)tempval >> -bits_free);
        bits_free += 32;
        ++m_pos;
        data[m_pos] = bits_free < 32 ? (tempval << bits_free) : 0;  // line 178: OOB WRITE
    }
    ...
}
```

The resize guard fires only when `len > bits_free` (strictly greater). When `len == bits_free` (the current slot is **exactly** filled), the guard is skipped. The code then enters the `bits_free <= 0` branch, increments `m_pos` to `data.size()` (one past the end), and writes to `data[m_pos]` — a heap buffer overflow.

### Trigger Condition

| Parameter | Value |
|---|---|
| `m_pos` at time of crash | `data.size() - 1` (last valid slot) |
| `bits_free` | `32` |
| `len` passed to `put_bits` | `32` |
| Resize guard `32 > 32` | **false** → no resize |
| After `bits_free -= len` | `bits_free = 0` |
| `bits_free <= 0` branch | entered → `++m_pos; data[m_pos] = 0` → **OOB WRITE** |

A `len = 32` call is reachable because `put_bits` is called with `cat` values from the Huffman DC coefficient path (line 1334–1335), where `cat` can equal 0, and separately from `put_val` which can encode symbols requiring up to 16 bits; the combination through repeated encoding of DCT blocks on a 1×1 frame exhausts the buffer quickly.

### Steps to reproduce

## PoC Code

```cpp
// Minimal PoC — no fuzzer infrastructure required
// Compile: see Reproduction Steps
#include <opencv4/opencv2/core.hpp>
#include <opencv4/opencv2/videoio.hpp>

int main() {
    uint8_t pixel = 0xff;
    cv::Mat frame(1, 1, CV_8UC1, &pixel);

    int fourcc = cv::VideoWriter::fourcc('M', 'J', 'P', 'G');
    cv::VideoWriter writer;
    writer.open("/tmp/poc_out.avi", 0, fourcc, 25.0, cv::Size(1, 1), false);

    if (writer.isOpened()) {
        writer.write(frame);   // triggers heap-buffer-overflow
    }
    return 0;
}
```

## Reproduction Steps

```bash
OUTPUT=<path-to-opencv-sanitizer-build>

clang++ -g -O0 -fstandalone-debug -fno-omit-frame-pointer \
    -fsanitize=address,undefined \
    -I$OUTPUT/include -I$OUTPUT/include/opencv4 \
    poc.cpp -o poc \
    -Wl,--start-group \
        $OUTPUT/lib/lib*.a \
        $OUTPUT/3rdparty/lib/lib*.a \
    -Wl,--end-group \
    -lz -ldl -lpthread -lm

./poc
```

## Stack Trace

```
[ WARN:0@0.003] global cap.cpp:778 open VIDEOIO(CV_IMAGES): raised OpenCV exception:

OpenCV(4.14.0-pre) /root/src/opencv/modules/videoio/src/cap_images.cpp:411: error: (-215:Assertion failed) !filename_pattern.empty() in function 'CvVideoWriter_Images'


=================================================================
==19802==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x78283bfe1050 at pc 0x5efefb5f46ab bp 0x7ffd077c7810 sp 0x7ffd077c7808
WRITE of size 4 at 0x78283bfe1050 thread T0
    #0 0x5efefb5f46aa in cv::mjpeg::mjpeg_buffer::put_bits(unsigned int, int) /root/src/opencv/modules/videoio/src/cap_mjpeg_encoder.cpp:178:25
    #1 0x5efefb5eebfd in cv::mjpeg::MjpegEncoder::operator()(cv::Range const&) const /root/src/opencv/modules/videoio/src/cap_mjpeg_encoder.cpp:1335:51
    #2 0x5efefb7317f6 in cv::parallel_for_impl(cv::Range const&, cv::ParallelLoopBody const&, double) /root/src/opencv/modules/core/src/parallel.cpp:626:5
    #3 0x5efefb731217 in cv::parallel_for_(cv::Range const&, cv::ParallelLoopBody const&, double) /root/src/opencv/modules/core/src/parallel.cpp:524:13
    #4 0x5efefb5eab2f in cv::mjpeg::MotionJpegWriter::writeFrameData(unsigned char const*, int, int, int) /root/src/opencv/modules/videoio/src/cap_mjpeg_encoder.cpp:1514:5
    #5 0x5efefb5f5f03 in cv::mjpeg::MotionJpegWriter::write(cv::_InputArray const&) /root/src/opencv/modules/videoio/src/cap_mjpeg_encoder.cpp:497:9
    #6 0x5efefb4a481e in cv::VideoWriter::write(cv::_InputArray const&) /root/src/opencv/modules/videoio/src/cap.cpp:871:18
    #7 0x5efefb493a02 in main /root/FuzzAgent/output/opencv/crash_reports/crash_003/poc.cpp:16:16

0x78283bfe1050 is located 0 bytes after 32-byte region [0x78283bfe1030,0x78283bfe1050)
allocated by thread T0 here:
    #0 0x5efefb492181 in operator new(unsigned long) (/root/FuzzAgent/output/opencv/crash_reports/crash_003/poc+0x723181) (BuildId: 11a449a353952b9ead9041da9dd29b1f519b5e65)
    #1 0x5efefb5f14ba in std::__new_allocator<unsigned int>::allocate(unsigned long, void const*) /usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/new_allocator.h:151:27
    #2 0x5efefb5f10cf in std::allocator_traits<std::allocator<unsigned int>>::allocate(std::allocator<unsigned int>&, unsigned long) /usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/alloc_traits.h:482:20
    #3 0x5efefb5f10cf in std::_Vector_base<unsigned int, std::allocator<unsigned int>>::_M_allocate(unsigned long) /usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/stl_vector.h:381:20
    #4 0x5efefb5f0b9c in std::vector<unsigned int, std::allocator<unsigned int>>::_M_default_append(unsigned long) /usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/vector.tcc:663:34
    #5 0x5efefb5f09d2 in std::vector<unsigned int, std::allocator<unsigned int>>::resize(unsigned long) /usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/stl_vector.h:1016:4
    #6 0x5efefb5ef52b in cv::mjpeg::mjpeg_buffer::resize(int) /root/src/opencv/modules/videoio/src/cap_mjpeg_encoder.cpp:158:14
    #7 0x5efefb5f428c in cv::mjpeg::mjpeg_buffer::put_bits(unsigned int, int) /root/src/opencv/modules/videoio/src/cap_mjpeg_encoder.cpp:166:13
    #8 0x5efefb5f4108 in cv::mjpeg::mjpeg_buffer::put_val(int, unsigned int const*) /root/src/opencv/modules/videoio/src/cap_mjpeg_encoder.cpp:189:9
    #9 0x5efefb5eebe5 in cv::mjpeg::MjpegEncoder::operator()(cv::Range const&) const /root/src/opencv/modules/videoio/src/cap_mjpeg_encoder.cpp:1334:51
    #10 0x5efefb7317f6 in cv::parallel_for_impl(cv::Range const&, cv::ParallelLoopBody const&, double) /root/src/opencv/modules/core/src/parallel.cpp:626:5
    #11 0x5efefb731217 in cv::parallel_for_(cv::Range const&, cv::ParallelLoopBody const&, double) /root/src/opencv/modules/core/src/parallel.cpp:524:13
    #12 0x5efefb5eab2f in cv::mjpeg::MotionJpegWriter::writeFrameData(unsigned char const*, int, int, int) /root/src/opencv/modules/videoio/src/cap_mjpeg_encoder.cpp:1514:5
    #13 0x5efefb5f5f03 in cv::mjpeg::MotionJpegWriter::write(cv::_InputArray const&) /root/src/opencv/modules/videoio/src/cap_mjpeg_encoder.cpp:497:9
    #14 0x5efefb4a481e in cv::VideoWriter::write(cv::_InputArray const&) /root/src/opencv/modules/videoio/src/cap.cpp:871:18
    #15 0x5efefb493a02 in main /root/FuzzAgent/output/opencv/crash_reports/crash_003/poc.cpp:16:16

SUMMARY: AddressSanitizer: heap-buffer-overflow
    cap_mjpeg_encoder.cpp:178:25 in cv::mjpeg::mjpeg_buffer::put_bits(unsigned int, int)
```

## Suggested Fix

Change the resize guard condition in `put_bits` from strict-greater-than to greater-than-or-equal, so that the buffer is expanded before the next slot is written when the current slot is exactly filled:

```diff
--- a/modules/videoio/src/cap_mjpeg_encoder.cpp
+++ b/modules/videoio/src/cap_mjpeg_encoder.cpp
@@ -163,7 +163,7 @@ class mjpeg_buffer
     inline void put_bits(unsigned bits, int len)
     {
         CV_Assert(len >=0 && len < 32);
-        if((m_pos == (data.size() - 1) && len > bits_free) || m_pos == data.size())
+        if((m_pos == (data.size() - 1) && len >= bits_free) || m_pos == data.size())
         {
             resize(int(2*data.size()));
         }
```

This ensures that when a `put_bits` call would exactly exhaust the last slot and then attempt to initialize the following one, the vector is resized first.

---
> Signed-off-by: FuzzAnything fuzzanything@gmail.com


### Issue submission checklist

- [x] I report the issue, it's not a question
- [x] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [x] I updated to the latest OpenCV version and the issue is still there
- [x] There is reproducer code and related data files (videos, images, onnx, etc)

--- a/modules/videoio/src/cap_mjpeg_encoder.cpp
+++ b/modules/videoio/src/cap_mjpeg_encoder.cpp
@@ -161,7 +161,7 @@ public:
     inline void put_bits(unsigned bits, int len)
     {
         CV_Assert(len >=0 && len < 32);
-        if((m_pos == (data.size() - 1) && len > bits_free) || m_pos == data.size())
+        if((m_pos == (data.size() - 1) && len >= bits_free) || m_pos == data.size())
         {
             resize(int(2*data.size()));
         }

--- a/modules/videoio/test/test_video_io.cpp
+++ b/modules/videoio/test/test_video_io.cpp
@@ -1280,4 +1280,22 @@ VideoCaptureAPIs seekable_backeinds[] = {CAP_FFMPEG, CAP_MSMF, CAP_AVFOUNDATION}
 
 INSTANTIATE_TEST_CASE_P(videoio, PreciseSeekingTest, testing::ValuesIn(seekable_backeinds), safe_capture_name_printer);
 
+// Regression test for heap-buffer-overflow in mjpeg_buffer::put_bits (GitHub issue #29112).
+// When len == bits_free the old guard used strict '>' and skipped the resize, causing
+// an out-of-bounds write after '++m_pos' advanced past data.size().
+TEST(Videoio_MJPEG, put_bits_no_heap_overflow)
+{
+    const std::string filename = cv::tempfile(".avi");
+    uint8_t pixel = 0xff;
+    cv::Mat frame(1, 1, CV_8UC1, &pixel);
+    int fourcc = cv::VideoWriter::fourcc('M', 'J', 'P', 'G');
+    {
+        cv::VideoWriter writer;
+        ASSERT_NO_THROW(writer.open(filename, CAP_OPENCV_MJPEG, fourcc, 25.0, cv::Size(1, 1), false));
+        ASSERT_TRUE(writer.isOpened());
+        EXPECT_NO_THROW(writer.write(frame));
+    }
+    remove(filename.c_str());
+}
+
 } // namespace

--- a/modules/videoio/test/test_video_io.cpp
+++ b/modules/videoio/test/test_video_io.cpp
@@ -1286,8 +1286,7 @@ INSTANTIATE_TEST_CASE_P(videoio, PreciseSeekingTest, testing::ValuesIn(seekable_
 TEST(Videoio_MJPEG, put_bits_no_heap_overflow)
 {
     const std::string filename = cv::tempfile(".avi");
-    uint8_t pixel = 0xff;
-    cv::Mat frame(1, 1, CV_8UC1, &pixel);
+    cv::Mat frame(1, 1, CV_8UC1, cv::Scalar::all(255));
     int fourcc = cv::VideoWriter::fourcc('M', 'J', 'P', 'G');
     {
         cv::VideoWriter writer;

